File: solutions/cpp/Dutch_national_flag.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::swap;
using std::uniform_int_distribution;
using std::vector;

// @include
void DutchFlagPartition(int pivot_index, vector<int>* A) {
  auto& A_ref = *A;
  int pivot = A_ref[pivot_index];
  /**
   * Keep the following invariants during partitioning:
   * bottom group: A_ref[0 : smaller - 1].
   * middle group: A_ref[smaller : equal - 1].
   * unclassified group: A_ref[equal : larger].
   * top group: A_ref[larger + 1 : A_ref.size() - 1].
   */
  int smaller = 0, equal = 0, larger = A_ref.size() - 1;
  // When there is any unclassified element.
  while (equal <= larger) {
    // A_ref[equal] is the incoming unclassified element.
    if (A_ref[equal] < pivot) {
      swap(A_ref[smaller++], A_ref[equal++]);
    } else if (A_ref[equal] == pivot) {
      ++equal;
    } else {  // A_ref[equal] > pivot.
      swap(A_ref[equal], A_ref[larger--]);
    }
  }
}
// @exclude

vector<int> RandVector(int len) {
  default_random_engine gen((random_device())());
  vector<int> ret;
  while (len--) {
    uniform_int_distribution<int> dis(0, 2);
    ret.emplace_back(dis(gen));
  }
  return ret;
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 100);
      n = dis(gen);
    }
    vector<int> A(RandVector(n));
    uniform_int_distribution<int> dis(0, A.size() - 1);
    int pivot_index = dis(gen);
    int pivot = A[pivot_index];
    DutchFlagPartition(pivot_index, &A);
    int i = 0;
    while (i < A.size() && A[i] < pivot) {
      cout << A[i] << ' ';
      ++i;
    }
    while (i < A.size() && A[i] < pivot) {
      cout << A[i] << ' ';
      ++i;
    }
    while (i < A.size() && A[i] < pivot) {
      cout << A[i] << ' ';
      ++i;
    }
    cout << endl;
    assert(i == A.size());
  }
  return 0;
}

File: solutions/python/dutch_national_flag.py
# Dutch_national_flag.cc
import random
import sys

# @include
def dutch_flag_partition(pivot_index, A):
    pivot = A[pivot_index]
    smaller, equal, larger = 0, 0, len(A) - 1
    while equal <= larger:
        if A[equal] < pivot:
            A[smaller], A[equal] = A[equal], A[smaller]
            smaller, equal = smaller + 1, equal + 1
        elif A[equal] == pivot:
            equal += 1
        else:
            A[equal], A[larger] = A[larger], A[equal]
            larger -= 1

# @exclude
def is_dutch_flag_sorted(A, pivot):
    count_increase = 0
    pivot_index = A.index(pivot)
    for smaller in A[:pivot_index]:
        if smaller > pivot:
            return False
    while A[pivot_index] == pivot:
        if pivot_index == len(A) - 1:
            return True
        pivot_index += 1
    for larger in A[pivot_index:len(A)]:
        if larger < pivot:
            return False
    return True

def main():
    for _ in range(1000):
        if len(sys.argv) == 2:
            n = int(argv[1])
        else:
            n = random.randint(1, 100)
        A = [random.randint(0, 10) for _ in range(n)]
        pivot_index = random.randint(0, len(A)-1)
        pivot = A[pivot_index]
        dutch_flag_partition(pivot_index, A)
        assert(is_dutch_flag_sorted(A, pivot))

if __name__=='__main__':
    main()
